consensus/state.go

```diff
@@ -1641,21 +1653,25 @@ func (cs *ConsensusState) addVote(vote *types.Vote, peerID p2p.ID) (added bool,
 	case types.VoteTypePrecommit:
 		precommits := cs.Votes.Precommits(vote.Round)
 		cs.Logger.Info("Added to precommit", "vote", vote, "precommits", precommits.StringShort())
+
 		blockID, ok := precommits.TwoThirdsMajority()
-		if ok && len(blockID.Hash) != 0 {
+		if ok {
 			// Executed as TwoThirdsMajority could be from a higher round
 			cs.enterNewRound(height, vote.Round)
 			cs.enterPrecommit(height, vote.Round)
-			cs.enterCommit(height, vote.Round)
-
-			if cs.config.SkipTimeoutCommit && precommits.HasAll() {
-				cs.enterNewRound(cs.Height, 0)
+			if len(blockID.Hash) != 0 {
+				cs.enterCommit(height, vote.Round)
+				if cs.config.SkipTimeoutCommit && precommits.HasAll() {
+					cs.enterNewRound(cs.Height, 0)
+				}
+			} else {
+				cs.enterPrecommitWait(height, vote.Round)
 			}
 		} else if cs.Round <= vote.Round && precommits.HasTwoThirdsAny() {
 			cs.enterNewRound(height, vote.Round)
-			cs.enterPrecommit(height, vote.Round)
```

Comment:
Jae was [suggesting](https://github.com/tendermint/tendermint/issues/1745#issuecomment-400858937) we also call `enterPropose` and `enterPrevoteWait` here. Note `enterNewRound` calls `enterPropose`, so it's not needed. And we want to wait the whole proposal step before doing any prevoting, so I don't think we need the `enterPrevoteWait` here either (it will happen latter if it needs to).

But then calling enterPrecommitWait right away means we now only have TimeoutPrecommit to receive the proposal block. Is that right?

In this case, we could be in any step, and may even be jumping to a later round,  but then we only give ourselves TimeoutPrecommit to receive the proposal block for that round.

Reply:
> But then calling enterPrecommitWait right away means we now only have TimeoutPrecommit to receive the proposal block. Is that right?
In this case, we could be in any step, and may even be jumping to a later round, but then we only give ourselves TimeoutPrecommit to receive the proposal block for that round.

That's why in Lemma 5 we need to assume that timeoutPropose(r) > 2Delta + timeoutPrecommit(r-1), i.e., timeoutPropose need to be big enough so we will wait for a process that jump to a higher round to receive Proposal. So we don't wait timeoutPrecommit to receive the proposal block. We wait timeoutPrecommit to set validBlock in the current round. And we wait timeoutPropose to receive the proposal, but there is dependency on the duration of timeoutPropose on timeoutPrecommit. Does this answer your question?